Add GovCloud Support

matt-corbalt opened the pull request:
## PR Description

Adds GovCloud support for building the AWS signin url when using the `kion console` command.

- adds an additional API call to CloudTamer to retrieve the account information to determine account type
- this change has been tested against Commercial and GovCloud accounts using `kion console`

## PR Checklist
* [ ] **New automated tests have been written to the extent possible.**
* [x] **The code has been checked for structural/syntactic validity.**
	- AMI/application: a build was performed
	- terraform changes: "terraform plan" checked on every affected environment
* [ ] **(If applicable) the code has been manually tested on our infrastructure.**
	- AMI/application: deployed an a test or dev environment
	- terraform changes: applied to test or dev environment
	- script: run against test or dev environment
* [x] **Likely failure points and new functionality have been identified and tested manually.**
	Examples:
	- Application manually run in a way that triggers any new branches
	- AMI logged into and changes verified from login shell
* [x] **Pull request description includes a description of all the manual steps performed to accomplish the above.**

To provide feedback on this template, visit https://docs.google.com/document/d/1YfTv7Amyop5G_8w1c2GJ_Mu-70L0KkZHhm9f9umDi3U/edit

matt-corbalt committed:
add govcloud support

hundt-corbalt commented on README.md:
```diff
@@ -27,6 +27,7 @@ app-api-key-duration: 168h0m0s
 host: kion.example.com
 idms: 1
 region: us-east-1
+govcloud-region: us-west-1
```
"us-gov-west-1"?

hundt-corbalt commented on cmd/kion/setup/setup.go:
```diff
@@ -194,11 +194,22 @@ func run() error {
 	}
 
+	var govcloudRegion string
+	err = survey.AskOne(
```
govcloud is such a niche feature that I'm not sure it makes sense to always ask about it in setup. what do you think @jonahb ?

jonahb replied:
I agree but I don't think we need this setting (see review comment).

hundt-corbalt commented on cmd/kion/console/console.go:
```diff
@@ -49,31 +50,52 @@ func run(cfg *config.Config, keyCfg *config.KeyConfig) error {
+
+	awsEndpoint, region := "", ""
+
+	// Account types: 1=commercial, 2=govcloud
```
use constants for account types and aws endpoints?

jonahb replied:
IMO, since the string literals for the AWS domains are only in one place, constants aren't needed.

jonahb commented on cmd/kion/console/console.go:
```diff
@@ -49,31 +50,52 @@ func run(cfg *config.Config, keyCfg *config.KeyConfig) error {
 		return err
 	}
+
+	awsEndpoint, region := "", ""
```
I think `awsDomain` would work better here. "Endpoint" is ambiguous.

jonahb reviewed (changes requested):
I would change the interface a bit. GovCloud regions are regions, so we don't need a separate setting — you can specify `--region us-gov-west-1`. If you want to save that for a certain account, you can put it in `kion.yml`, which overrides the global region. (Unless I'm missing a scenario?)

We need to know the AWS domain given a region (`aws.amazon.com` or `amazonaws-us-gov.com`). I'd rather not call into Kion for this info. That seems brittle, and it adds an extra call. Instead, it seems safe enough to do:

```
awsDomain := strings.HasPrefix(region, "us-gov-") ? "amazonaws-us-gov.com" : "aws.amazon.com"
```

jonahb commented:
> This makes sense generally, but for my use-case I often don't have the upfront information that states whether or not an account ID is tied to GovCloud or Commercial (I can explain this further if that's helpful).

@matt-corbalt Could you explain further? I think this would help us get to the right design. (I'm probably using the tool in a different way since I deal with a fixed set of accounts.)

jonahb committed:
Remove region

matt-corbalt commented:
@jonahb I rebased your `jonah/no-region` changes on top of my branch. As requested, I left the GovCloud detection and removed the GovCloud region argument.

The console cmd still works as expected! Let me know how you'd like to proceed.

jonahb reviewed (changes requested):
This is looking really good. I left a few more comments about cleaning up the `client.Client` interface.

matt-corbalt commented on internal/client/client.go:
```diff
@@ -48,6 +48,15 @@ type TemporaryCredentials struct {
+	AccountName   string `json:"account_name"`
+	AccountNumber string `json:"account_number"`
+	AccountTypeID int64  `json:"account_type_id"`
+	CreatedAt     string `json:"created_at"`
```
Yep, agreed. I choose to keep those fields since they might be useful in the future for other use-cases, but definitely not needed here and better to omit them to make this cleaner.

matt-corbalt committed:
Improve code readability

matt-corbalt commented on internal/client/client.go:
```diff
@@ -48,6 +48,15 @@ type TemporaryCredentials struct {
+type AccountDetails struct {
+	AccountName   string `json:"account_name"`
+	AccountNumber string `json:"account_number"`
+	AccountTypeID int64  `json:"account_type_id"`
```
I ended up adding this and modifying the domain switch statement when it's used to the following:

```
var awsDomain string
switch accountInfo.Type {
case client.AccountTypeCommercial:
	awsDomain = "aws.amazon.com"
case client.AccountTypeGovCloud:
	awsDomain = "amazonaws-us-gov.com"
default:
	panic(fmt.Sprintf("unexpected account type: %d", accountInfo.Type))
}
```

I'm not a Go expert, it this a reasonable approach?

jonahb replied:
The switch is reasonable. I don't think we should panic if we don't get AWS Commercial or GovCloud though. Better to return an error and let the program exit in the normal way.

jonahb commented on internal/client/client.go:
```diff
@@ -48,6 +48,20 @@ type TemporaryCredentials struct {
+type AccountType int
+
+const (
+	AccountTypeUnknown    AccountType = 0
```
Geez, on second thought, `AccountTypeUnknown` isn't necessary. I was thinking in terms of languages with a fixed set of "enum" values.

jonahb commented on cmd/kion/console/console.go:
```diff
@@ -96,8 +108,7 @@ func run(cfg *config.Config, keyCfg *config.KeyConfig) error {
 }
-
-func getAWSSigninToken(accessKeyID string, secretAccessKey string, sessionToken string) (string, error) {
+func getAWSSigninToken(awsEndpoint string, accessKeyID string, secretAccessKey string, sessionToken string) (string, error) {
```
Use `awsDomain` as the parameter as well?

jonahb reviewed (changes requested):
Looks good. A few tweaks and this is ready to merge.

matt-corbalt committed:
Return an error instead of panic-ing when determining aws domain from account type

matt-corbalt committed:
Remove AccountTypeUnknown

matt-corbalt committed:
Use consistent variable names for awsDomain

matt-corbalt commented:
@jonahb Sorry for the delay. Updated the code per your feedback. Let me know if you see anything else!

jonahb reviewed (approved):
Looks great!
